juju/main.py

```diff
@@ -0,0 +1,218 @@
+    except AttributeError:
+        debugf(f"event {formatted_event_name} not defined for {charm}")
+
+    # If the event is not supported by the charm implementation, do
+    # not error out or try to emit it. This is to support rollbacks.
+    if event_to_emit is not None:
+        debugf(f'Emitting Juju event {event_name}')
+        event_to_emit.emit()
+
+
+def main():
+    """An entry point for a charm.
+
+    Sets up in-memory state (object hierarchy and observation etc.).
+
+    The expected hierarchy is a single charm object under the framework
+    object with many child objects under the charm object:
+
+    framework -> charm object -> member child objects (endpoints, components)
+
+    The charm class should hold all necessary setup code for the modeled
+    application such as object instantiation, setup of observation of events
+    (which may also be done in child classes).
+
+    """
+    charm_module = importlib.import_module('charm')
+    juju_charm_module = importlib.import_module('juju.charm')
+    framework_module = importlib.import_module('juju.framework')
+
+    charm_dir = _get_charm_dir()
+    juju_event_name = os.path.basename(sys.argv[0])
+    # Set up hooks on initial install or charm upgrade. Note that if
+    # a charm is force-upgraded this will only be called after
+    # upgrade-charm is executed.
+    if juju_event_name in ['install', 'upgrade-charm']:
+        _setup_hooks(charm_dir, charm_module.Charm,
+                     juju_charm_module.CharmEvents,
+                     framework_module.Event, framework_module.BoundEvent)
+
+    # TODO: If Juju unit agent crashes after exit(0) from the charm code
+    # the framework will commit the snapshot but Juju will not commit its
+    # operation.
+    framework = None
+    try:
+        charm_state_path = os.path.join(charm_dir, CHARM_STATE_FILE)
+        framework = framework_module.Framework(data_path=charm_state_path)
+
+        # Set up charm-specific in-memory state.
+        charm = charm_module.Charm(framework, None)
+
+        framework.reemit()
+
+        # Process the Juju event relevant to the current hook execution
+        # JUJU_HOOK_NAME or JUJU_ACTION_NAME are not used to support simulation
+        # of events from debugging sessions.
+        _emit_charm_event(charm, juju_event_name)
+
+        framework.commit()
+    finally:
+        if framework:
+            framework.close()
+
+
+def setup_path():
+    # Use modules from the $JUJU_CHARM_DIR/lib directory (conditionally, to
+    # avoid side-effects on module import).
+    # An absolute path is used as using a symlink like
+    # hooks/install -> ../lib/juju/main.py
+    # will result in $JUJU_CHARM_DIR/lib/juju being the first path stored
+    # in sys.path during hook execution which is undesirable.
+    # This happens because the directory containing a script is added to
+    # sys.path by the interpreter which gets resolved if it is a symlink.
```

Comment:
I don't quite get the comment above. First, the absolute path is used because it sounds like the right thing to do. If it's relative, it will be relative whereved the current process decided to move over to. And then, it sounds like the comment above is also saying that the directory containing the executable is only being added to the path if it's a symlink, and that wouldn't happen if it was an absolute path, but that's not true I think. I think this comment should have a form closer to:
```
# First element is the directory containing this file. We don't want that in the path.
del ...
# Then, add $JUJU_CHARM_DIR/lib to the path.
... insert ...
```
And done?

Reply:
Agreed, I think we can have a much shorter comment.